components/dashboard/src/workspaces/CreateWorkspacePage.tsx

```diff
@@ -0,0 +1,346 @@
+            : user?.additionalData?.ideSettings?.defaultIde,
+    );
+    const [selectedWsClass, setSelectedWsClass] = useState<string | undefined>(props.workspaceClass);
+    const [errorWsClass, setErrorWsClass] = useState<string | undefined>(undefined);
+    const [repo, setRepo] = useState<string | undefined>(location.hash.substring(1));
+    const onSelectEditorChange = useCallback(
+        (ide: string, useLatest: boolean) => {
+            setSelectedIde(ide);
+            setUseLatestIde(useLatest);
+        },
+        [setSelectedIde, setUseLatestIde],
+    );
+    const [errorIde, setErrorIde] = useState<string | undefined>(undefined);
+    const [creating, setCreating] = useState<boolean>(false);
+
+    const [createResult, setCreateResult] = useState<WorkspaceCreationResult | undefined>(undefined);
+    const [createError, setCreateError] = useState<StartWorkspaceError | undefined>(undefined);
+    const [selectAccountError, setSelectAccountError] = useState<SelectAccountPayload | undefined>(undefined);
+
+    const createWorkspace = useCallback(
+        async (options?: Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">) => {
+            // add options from search params
+            const opts = options || {};
+
+            if (!opts.workspaceClass) {
+                opts.workspaceClass = selectedWsClass;
+            }
+            if (!opts.ideSettings) {
+                opts.ideSettings = {
+                    defaultIde: selectedIde,
+                    useLatestVersion: useLatestIde,
+                };
+            }
+            if (!repo) {
+                return;
+            }
+
+            try {
+                setCreating(true);
+                const result = await getGitpodService().server.createWorkspace({
+                    contextUrl: repo,
+                    ...opts,
+                });
+
+                setCreateResult(result);
+            } catch (error) {
+                setCreateError(error);
+            } finally {
+                setCreating(false);
+            }
+        },
+        [repo, selectedIde, selectedWsClass, useLatestIde],
+    );
+
+    const onClickCreate = useCallback(() => createWorkspace(), [createWorkspace]);
+
+    if (SelectAccountPayload.is(selectAccountError)) {
+        return (
+            <SelectAccountModal
+                {...selectAccountError}
+                close={() => {
+                    window.location.href = gitpodHostUrl.asAccessControl().toString();
+                }}
+            />
+        );
+    }
+    if (createResult?.workspaceURL) {
+        window.location.href = createResult.workspaceURL;
+        return <></>;
+    } else if (createResult?.createdWorkspaceId) {
+        return <Redirect to={`/start/${createResult.createdWorkspaceId}`} />;
+    } else if (createResult?.existingWorkspaces) {
+        return <ExistingWorkspaceModal result={createResult} createWorkspace={createWorkspace} onClose={() => {}} />;
```

Comment:
Instead of relying on the result for this check here, I think it would be clearer if we had something like a `showExistingWorkspaceModal` component state, and we just set that in the `createWorkspace()` function if this condition happens, and render the modal based only on that boolean flag.

Reply:
I was wrong here, it needs the result to render, more details in my response above.